README.md
# Bomb-Evader

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## Unreleased

## [1.0.0] - 2021-04-06
### Added
- Create .editorconfig to allow for consistent code formatting.
- Added changelog to actively check changes to codebase.
- Create config file for the game.
- Add webpack config files.
- Create the scenes for the game. 
- Add Leaderboard access.
- Create the player object.
- Style the game

src/utils/api.js
export default class Api {
  static get baseUrl() {
    return 'https://us-central1-js-capstone-backend.cloudfunctions.net/api/games';
  }

  static get gameId() {
    return 'OZBsTg9qC5go43D8PLU6';
  }

  static getBoard() {
    return Api.fetchData(`${Api.baseUrl}/${Api.gameId}/scores`).then(res => res);
  }

  static addScore() {
    return Api.fetchData(`${Api.baseUrl}/${Api.gameId}/scores`, 'POST', {
      user: window.localStorage.getItem('name') || 'No one',
      score: window.localStorage.getItem('score') || 0,
    }).then(res => res);
  }

  static async fetchData(url = '', method = 'GET', data = {}) {
    const options = {
      method,
      mode: 'cors', // no-cors, *cors, same-origin
      headers: {
        'Content-Type': 'application/json',
      },
    };

    if (method === 'POST') {
      options.body = JSON.stringify(data);
    }

    const response = await fetch(url, options);
    return response.json();
  }
}

webpack.config.js
const path = require('path');
const CopyWebpackPlugin = require('copy-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const webpack = require('webpack');
const { CleanWebpackPlugin } = require('clean-webpack-plugin');

module.exports = {
  entry: {
    app: './src/index.js',
    dependencies: ['phaser'],
  },
  output: {
    path: path.resolve(__dirname, 'build'),
    filename: '[name].bundle.js',
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        include: path.resolve(__dirname, 'src/'),
        loader: 'babel-loader',
      },
      {
        test: /\.css$/,
        exclude: /node_modules/,
        loader: ['style-loader', 'css-loader'],
      },
    ],
  },
  plugins: [
    new CopyWebpackPlugin({
      patterns: [
        {
          from: path.resolve(__dirname, 'assets'),
          to: path.resolve(__dirname, 'build', 'assets'),
        },
      ],
    }),
    new webpack.DefinePlugin({
      'typeof CANVAS_RENDERER': JSON.stringify(true),
      'typeof WEBGL_RENDERER': JSON.stringify(true),
    }),
    new HtmlWebpackPlugin({
      template: path.resolve(__dirname, 'index.html'),
    }),
    new CleanWebpackPlugin(),
  ],
  devServer: {
    contentBase: path.resolve(__dirname, 'build'),
  },
};

src/objects/player.js
import { Physics, Math } from 'phaser';

export default class Player extends Physics.Arcade.Sprite {
  constructor(scene) {
    super(scene, 300, 650, 'player');
    this.scene = scene;
    this.scene.add.existing(this);
    this.scene.physics.add.existing(this);
    this.cursors = this.scene.input.keyboard.createCursorKeys();
    this.setCollideWorldBounds(true);
    this.registerAnimations();
  }

  registerAnimations() {
    this.scene.anims.create({
      key: 'left',
      frames: this.scene.anims.generateFrameNumbers('player', { start: 0, end: 3 }),
      frameRate: 10,
      repeat: -1,
    });

    this.scene.anims.create({
      key: 'turn',
      frames: [{ key: 'player', frame: 4 }],
      frameRate: 20,
    });

    this.scene.anims.create({
      key: 'right',
      frames: this.scene.anims.generateFrameNumbers('player', { start: 5, end: 8 }),
      frameRate: 10,
      repeat: -1,
    });
  }

  static get xVelocity() {
    return 300;
  }

  static get yVelocity() {
    return 800;
  }

  moveLeft() {
    this.setVelocityX(-Player.xVelocity);
    this.anims.play('left', true);
  }

  moveRight() {
    this.setVelocityX(Player.xVelocity);
    this.anims.play('right', true);
  }

  standStill() {
    this.setVelocityX(0);
    this.anims.play('turn', true);
  }

  jump() {
    this.setVelocityY(-Player.yVelocity);
  }

  collectStar(player, star) {
    star.disableBody(true, true);

    this.score += 10;
    this.scoreText.setText(`Score: ${this.score}`);

    if (this.stars.countActive(true) === 0) {
      this.stars.children.iterate((child) => {
        child.enableBody(true, child.x, 0, true, true);
      });

      const x = (this.x < 400) ? Math.Between(400, 800) : Math.Between(0, 400);

      const bomb = this.bombs.create(x, 16, 'bomb');
      bomb.setBounce(1);
      bomb.setCollideWorldBounds(true);
      bomb.setVelocity(Math.Between(-200, 200), 20);
      bomb.allowGravity = false;
    }
  }

  hitBomb() {
    this.scene.physics.pause();

    this.setTint(0xff0000);

    this.anims.play('turn');

    this.scene.gameOver = true;
    this.scene.gameOverText.setVisible(true);
  }
}

src/scenes/leaderboard.js
import { Scene } from 'phaser';
import Api from '../utils/api';

export default class LeaderBoard extends Scene {
  constructor() {
    super('LeaderBoard');
  }

  // eslint-disable-next-line
  preload() {
    if (!window.localStorage.getItem('name')) {
      window.localStorage.setItem('name', window.prompt('Please provide a name: '));
    }
  }

  create() {
    this.title = this.add.text(600, 150, 'Leader Board', { fontSize: '48px', fill: '#fff' }).setOrigin(0.5);
    this.wait = this.add.text(600, 350, 'Please Wait...', { fontSize: '32px', fill: '#fff' }).setOrigin(0.5);

    Api.addScore().then(() => {
      Api.getBoard().then(res => {
        this.wait.setVisible(false);
        res = res.result;
        let point = 350;
        let order = 1;
        res.sort((a, b) => ((parseInt(a.score, 10) < parseInt(b.score, 10)) ? 1 : -1));

        res.slice(0, 10).forEach(result => {
          this.add.text(600, point, `(${order}) - ${result.user} : ${result.score}`).setOrigin(0.5);
          point += 20;
          order += 1;
        });

        this.add.text(600, point + 30, 'Press any key...', { fontSize: '32px', fill: '#fff' }).setOrigin(0.5);
        window.localStorage.removeItem('name');
      });
    }).finally(() => {
      this.input.keyboard.on('keydown', () => {
        this.scene.start('Main');
      });
    });
  }
}

src/scenes/main.js
import { Scene, Math } from 'phaser';
import Player from '../objects/player';

export default class Main extends Scene {
  constructor() {
    super('Main');
    this.gameOver = false;
  }

  preload() {
    this.load.image('sky', 'assets/bg.png');
    this.load.image('ground', 'assets/platform.png');
    this.load.image('star', 'assets/star.png');
    this.load.image('bomb', 'assets/bomb.png');
    this.load.spritesheet('player', 'assets/dude.png', { frameWidth: 32, frameHeight: 48 });
  }

  create() {
    this.score = 0;
    window.localStorage.setItem('score', this.score);
    this.scoreText = this.add.text(16, 16, 'score: 0', { fontSize: '32px', fill: '#fff' });
    this.gameOverText = this.add.text(600, 250, 'Game Over (Press any key)', { fontSize: '48px', fill: '#fff' }).setVisible(false).setDepth(2).setOrigin(0.5);

    this.bg = this.add.tileSprite(640, 374, 1280, 768, 'sky').setDepth(-1);

    this.platforms = this.physics.add.staticGroup();

    this.platforms.create(300, 200, 'ground');
    this.platforms.create(1010, 350, 'ground');
    this.platforms.create(375, 525, 'ground');
    this.platforms.create(600, 750, 'ground').setScale(3.5).refreshBody();

    this.player = new Player(this);
    this.cursors = this.input.keyboard.createCursorKeys();

    this.stars = this.physics.add.group({
      key: 'star',
      repeat: 15,
      setXY: { x: 12, y: 0, stepX: 83 },
    });

    this.stars.children.iterate((child) => {
      child.setBounceY(Math.FloatBetween(0.4, 0.8));
    });

    this.bombs = this.physics.add.group();

    this.physics.add.overlap(this.player, this.stars, this.player.collectStar, null, this);

    this.physics.add.collider(this.player, this.bombs, this.player.hitBomb, null, this.player);

    this.physics.add.collider(this.player, this.platforms);
    this.physics.add.collider(this.stars, this.platforms);
    this.physics.add.collider(this.bombs, this.platforms);

    this.input.keyboard.on('keydown', () => {
      if (this.gameOver) {
        this.gameOver = false;
        window.localStorage.setItem('score', this.score);
        this.scene.start('LeaderBoard');
      }
    });
  }

  update() {
    if (this.gameOver) {
      return;
    }

    if (this.cursors.left.isDown) {
      this.player.moveLeft();
    } else if (this.cursors.right.isDown) {
      this.player.moveRight();
    } else {
      this.player.standStill();
    }

    if (this.cursors.up.isDown && this.player.body.touching.down) {
      this.player.jump();
    }
  }
}